File: src/vault/queue.ts
/**
 * File-based queue for Phase 0 brain synthesis.
 *
 * Queue items live in <vault>/_queue/pending/<timestamp>-<nanoid>.json.
 * Claiming an item atomically renames .json -> .claimed so a single
 * synthesize call processes it. Done items move to _queue/done/.
 *
 * No external lock service: the rename itself is the lock. Two concurrent
 * claimNextItem() calls cannot both rename the same file — one wins, the
 * other gets ENOENT and skips to the next pending entry.
 */
import fs from 'node:fs/promises';
import path from 'node:path';
import { randomBytes } from 'node:crypto';
import { CONFIG } from '../config.js';
import { writeAtomicFile } from './filesystem.js';
import { logger } from '../shared/logger.js';

export interface QueueItem {
  raw_path?: string;        // relative to vault; absent for deferred-fetch items
  source: 'curated' | 'gathered';
  captured_at: string;      // ISO
  title: string;
  source_url?: string;
  format: 'markdown' | 'html' | 'text' | 'pdf';
  content_hash?: string;    // SHA256 hex; absent for deferred-fetch items
  deferred_fetch?: boolean; // if true, brain_synthesize fetches source_url before processing
}

function pendingDir(): string {
  return path.join(CONFIG.VAULT_PATH, CONFIG.QUEUE_PENDING);
}

function doneDir(): string {
  return path.join(CONFIG.VAULT_PATH, CONFIG.QUEUE_DONE);
}

function nanoid(): string {
  return randomBytes(6).toString('hex');
}

/**
 * Write a queue item to _queue/pending/<timestamp>-<nanoid>.json.
 * Returns the absolute path of the written file.
 */
export async function enqueueItem(item: QueueItem): Promise<string> {
  await fs.mkdir(pendingDir(), { recursive: true });
  // Use compact ISO timestamp so lexicographic ordering matches chronological
  const stamp = new Date().toISOString().replace(/[:.]/g, '-');
  const name = `${stamp}-${nanoid()}.json`;
  const filePath = path.join(pendingDir(), name);
  await writeAtomicFile(filePath, JSON.stringify(item, null, 2) + '\n');
  logger.debug('Enqueued item', { name, raw_path: item.raw_path });
  return filePath;
}

/**
 * Claim the next pending item atomically by renaming .json -> .claimed.
 * Returns [claimedPath, item] or null if the queue is empty.
 *
 * Concurrent callers race on rename(); the loser sees ENOENT and tries the
 * next candidate, so multiple synthesizers can drain the queue safely.
 */
export async function claimNextItem(): Promise<[string, QueueItem] | null> {
  const dir = pendingDir();
  let entries: string[];
  try {
    entries = await fs.readdir(dir);
  } catch {
    return null;
  }

  // Sort so the oldest pending item is tried first.
  const pending = entries.filter((n) => n.endsWith('.json')).sort();

  for (const name of pending) {
    const fromPath = path.join(dir, name);
    const toPath = path.join(dir, name.replace(/\.json$/, '.claimed'));
    try {
      await fs.rename(fromPath, toPath);
    } catch {
      // Another worker grabbed this one (ENOENT) — try the next candidate.
      continue;
    }
    try {
      const raw = await fs.readFile(toPath, 'utf-8');
      const item = JSON.parse(raw) as QueueItem;
      return [toPath, item];
    } catch (err) {
      // The claim file is corrupt. Move it out of the way so we don't loop on it.
      logger.warn('Corrupt queue item; moving to done as poison pill', { name, error: String(err) });
      try {
        await fs.mkdir(doneDir(), { recursive: true });
        await fs.rename(toPath, path.join(doneDir(), `${name}.corrupt`));
      } catch { /* best-effort */ }
      continue;
    }
  }

  return null;
}

/**
 * Mark a claimed item as done by moving it into _queue/done/.
 */
export async function markDone(claimedPath: string): Promise<void> {
  const name = path.basename(claimedPath).replace(/\.claimed$/, '.done.json');
  const target = path.join(doneDir(), name);
  await fs.mkdir(doneDir(), { recursive: true });
  await fs.rename(claimedPath, target);
  logger.debug('Marked queue item done', { name });
}

/**
 * Unclaim a previously claimed item by renaming .claimed back to .json,
 * making it visible to future claimNextItem() calls. Used when synthesis
 * fails after a successful claim.
 */
export async function unclaimItem(claimedPath: string): Promise<void> {
  const restored = claimedPath.replace(/\.claimed$/, '.json');
  await fs.rename(claimedPath, restored);
  logger.debug('Unclaimed queue item', { path: restored });
}

File: internal/server/queue.go
package server

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// QueueItem is the JSON file the reaper drops into queue/pending/
// for each newly-arrived raw document. Mirrors the TS QueueItem
// shape (src/vault/queue.ts) so a same-vault TS synthesizer running
// in parallel could still claim items. Only fields actively used
// in Phase 2 are populated; the others stay omitted (omitempty)
// rather than zero so the JSON is small + diff-friendly.
type QueueItem struct {
	RawPath     string `json:"raw_path"`               // Raw/{curated,gathered}/<file>
	Source      string `json:"source"`                 // "curated" | "gathered"
	CapturedAt  string `json:"captured_at"`            // RFC3339
	Title       string `json:"title,omitempty"`
	Format      string `json:"format,omitempty"`       // "markdown" | "html" | ...
	ContentHash string `json:"content_hash,omitempty"` // SHA256 hex
	SourceURL   string `json:"source_url,omitempty"`
	SourceAttachment string `json:"source_attachment,omitempty"` // Raw/attachments/<sha>.ext
}

// EnqueueItem writes a QueueItem to queue/pending/<timestamp>-<rand>.json
// using vault.WriteAtomicFile semantics so a synthesizer claiming
// items concurrently cannot see a half-written file.
//
// The filename pattern guarantees per-process uniqueness via the
// 12-hex suffix (collision odds are negligible for the volumes we
// expect). Same-name collision across two daemons is impossible
// since the daemon owns the global flock.
func EnqueueItem(dataDir DataDir, profile, vaultName string, item QueueItem, suffix string) (string, error) {
	if item.RawPath == "" {
		return "", errors.New("server: EnqueueItem requires raw_path")
	}
	if item.Source == "" {
		return "", errors.New("server: EnqueueItem requires source")
	}
	if item.CapturedAt == "" {
		item.CapturedAt = time.Now().UTC().Format(time.RFC3339)
	}
	pendingDir := filepath.Join(dataDir.VaultDir(profile, vaultName), "queue", "pending")
	if err := os.MkdirAll(pendingDir, 0o755); err != nil {
		return "", fmt.Errorf("server: mkdir queue/pending: %w", err)
	}
	stamp := time.Now().UTC().Format("20060102T150405.000Z")
	name := fmt.Sprintf("%s-%s.json", stamp, suffix)
	path := filepath.Join(pendingDir, name)
	body, err := json.MarshalIndent(item, "", "  ")
	if err != nil {
		return "", err
	}
	if err := os.WriteFile(path, append(body, '\n'), 0o644); err != nil {
		return "", err
	}
	return path, nil
}

// ListPending returns every pending queue item path, sorted by name
// (which sorts chronologically since the prefix is a timestamp).
// Returns nil + nil for an empty/missing dir.
func ListPending(dataDir DataDir, profile, vaultName string) ([]string, error) {
	pendingDir := filepath.Join(dataDir.VaultDir(profile, vaultName), "queue", "pending")
	entries, err := os.ReadDir(pendingDir)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var out []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		if !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		out = append(out, filepath.Join(pendingDir, e.Name()))
	}
	sort.Strings(out)
	return out, nil
}

// ClaimNextItem moves the oldest pending item to queue/claimed/ via
// atomic rename — the rename IS the lock, so two concurrent claimers
// cannot grab the same file. Returns (claimedPath, item, nil) on
// success, (\"\", nil, nil) when the queue is empty.
//
// Phase 2 synthesizer claims items one at a time; the per-vault
// runner's mutex keeps the claim sequenced with reaper ingests, but
// the rename atomicity is what protects against external claimers
// (e.g. a same-vault TS server still running).
func ClaimNextItem(dataDir DataDir, profile, vaultName string) (string, *QueueItem, error) {
	pending, err := ListPending(dataDir, profile, vaultName)
	if err != nil {
		return "", nil, err
	}
	if len(pending) == 0 {
		return "", nil, nil
	}
	claimedDir := filepath.Join(dataDir.VaultDir(profile, vaultName), "queue", "claimed")
	if err := os.MkdirAll(claimedDir, 0o755); err != nil {
		return "", nil, err
	}
	for _, src := range pending {
		dst := filepath.Join(claimedDir, filepath.Base(src))
		if err := os.Rename(src, dst); err != nil {
			if errors.Is(err, os.ErrNotExist) {
				// Another claimer beat us; try the next one.
				continue
			}
			return "", nil, fmt.Errorf("server: claim rename: %w", err)
		}
		raw, rerr := os.ReadFile(dst)
		if rerr != nil {
			return "", nil, fmt.Errorf("server: read claimed: %w", rerr)
		}
		var item QueueItem
		if err := json.Unmarshal(raw, &item); err != nil {
			return "", nil, fmt.Errorf("server: parse %s: %w", dst, err)
		}
		return dst, &item, nil
	}
	return "", nil, nil
}

// MarkDone moves a claimed item to queue/done/. Called once the
// synthesizer has successfully written summary + entities + log.
// Idempotent — a double-mark just renames a non-existent file and
// returns os.ErrNotExist, which the caller can ignore.
func MarkDone(dataDir DataDir, profile, vaultName, claimedPath string) error {
	doneDir := filepath.Join(dataDir.VaultDir(profile, vaultName), "queue", "done")
	if err := os.MkdirAll(doneDir, 0o755); err != nil {
		return err
	}
	dst := filepath.Join(doneDir, filepath.Base(claimedPath))
	return os.Rename(claimedPath, dst)
}

// MarkDead moves a claimed item to queue/dead/ when synthesis failed
// in a way the synthesizer doesn't want to retry. New in Phase 2 —
// the TS implementation silently re-queued, which masked persistent
// gate failures behind apparent retries.
func MarkDead(dataDir DataDir, profile, vaultName, claimedPath string) error {
	deadDir := filepath.Join(dataDir.VaultDir(profile, vaultName), "queue", "dead")
	if err := os.MkdirAll(deadDir, 0o755); err != nil {
		return err
	}
	dst := filepath.Join(deadDir, filepath.Base(claimedPath))
	return os.Rename(claimedPath, dst)
}
